File: master/tests/test_protocol.py
"""Tests for tmon.protocol."""

import random

import pytest

from tmon.protocol import (
    PROTO_CMD_POLL,
    PROTO_CMD_REPLY,
    PROTO_START,
    PROTO_REPLY_PAYLOAD_LEN,
    PROTO_TEMP_INVALID,
    crc16_modbus,
    decode_frame,
    encode_request,
    parse_reply,
)


# -- CRC known vectors from protocol.org -------------------------------------


class TestCrc16Modbus:
    """CRC-16/MODBUS computation tests."""

    def test_example1_poll_slave3(self):
        """CRC of [03 01 00] should be 0x5080 (Example 1 in spec)."""
        assert crc16_modbus(bytes([0x03, 0x01, 0x00])) == 0x5080

    def test_example2_reply_slave3(self):
        """CRC of Example 2 body should be 0xEB90 (spec)."""
        body = bytes([
            0x03, 0x02, 0x08,
            0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
        ])
        assert crc16_modbus(body) == 0xEB90

    def test_empty_input(self):
        """CRC of empty data should be the initial value 0xFFFF."""
        assert crc16_modbus(b"") == 0xFFFF

    def test_single_byte(self):
        """CRC of a single zero byte should not be 0xFFFF."""
        result = crc16_modbus(b"\x00")
        assert result != 0xFFFF
        assert 0 <= result <= 0xFFFF


# -- encode_request ----------------------------------------------------------


class TestEncodeRequest:
    """Tests for the general frame encoder."""

    def test_example1_poll_slave3(self):
        """POLL for slave 3 should produce the Example 1 frame."""
        expected = bytes([0x01, 0x03, 0x01, 0x00, 0x80, 0x50])
        assert encode_request(3, PROTO_CMD_POLL, b"") == expected

    def test_poll_length_is_6(self):
        """POLL frame with no payload is always 6 bytes."""
        assert len(encode_request(1, PROTO_CMD_POLL, b"")) == 6
        assert len(encode_request(247, PROTO_CMD_POLL, b"")) == 6

    def test_example2_reply_frame(self):
        """encode_request should produce the Example 2 frame."""
        payload = bytes([
            0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
        ])
        expected = bytes([
            0x01, 0x03, 0x02, 0x08,
            0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
            0x90, 0xEB,
        ])
        assert encode_request(3, PROTO_CMD_REPLY, payload) == expected

    def test_start_byte(self):
        """First byte of any frame is PROTO_START."""
        frame = encode_request(1, PROTO_CMD_POLL, b"")
        assert frame[0] == PROTO_START

    def test_addr_in_frame(self):
        """Address appears at offset 1."""
        frame = encode_request(42, PROTO_CMD_POLL, b"")
        assert frame[1] == 42

    def test_cmd_in_frame(self):
        """Command byte appears at offset 2."""
        frame = encode_request(1, 0xAB, b"")
        assert frame[2] == 0xAB

    def test_len_field(self):
        """LEN field at offset 3 reflects actual payload length."""
        frame = encode_request(1, PROTO_CMD_POLL, b"\x01\x02\x03")
        assert frame[3] == 3

    def test_addr_boundary_low(self):
        """Address 0 should be rejected."""
        with pytest.raises(ValueError):
            encode_request(0, PROTO_CMD_POLL, b"")

    def test_addr_boundary_high(self):
        """Address 248 should be rejected."""
        with pytest.raises(ValueError):
            encode_request(248, PROTO_CMD_POLL, b"")


# -- decode_frame ------------------------------------------------------------


class TestDecodeFrame:
    """Tests for frame decoding and validation."""

    def test_roundtrip_poll(self):
        """encode then decode should recover the original fields."""
        raw = encode_request(5, PROTO_CMD_POLL, b"")
        frame = decode_frame(raw)
        assert frame["addr"] == 5
        assert frame["cmd"] == PROTO_CMD_POLL
        assert frame["payload"] == b""

    def test_roundtrip_reply(self):
        """Round-trip with a non-empty payload."""
        payload = bytes([0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F])
        raw = encode_request(10, PROTO_CMD_REPLY, payload)
        frame = decode_frame(raw)
        assert frame["addr"] == 10
        assert frame["cmd"] == PROTO_CMD_REPLY
        assert frame["payload"] == payload

    def test_example1_from_spec(self):
        """Decode the Example 1 frame from the protocol spec."""
        raw = bytes([0x01, 0x03, 0x01, 0x00, 0x80, 0x50])
        frame = decode_frame(raw)
        assert frame["addr"] == 3
        assert frame["cmd"] == PROTO_CMD_POLL
        assert frame["payload"] == b""

    def test_example2_from_spec(self):
        """Decode the Example 2 frame from the protocol spec."""
        raw = bytes([
            0x01, 0x03, 0x02, 0x08,
            0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
            0x90, 0xEB,
        ])
        frame = decode_frame(raw)
        assert frame["addr"] == 3
        assert frame["cmd"] == PROTO_CMD_REPLY
        assert len(frame["payload"]) == 8

    def test_error_short_frame(self):
        """Frames shorter than 6 bytes should be rejected."""
        with pytest.raises(ValueError, match="too short"):
            decode_frame(b"\x01\x02\x03")

    def test_error_bad_start(self):
        """Non-0x01 START byte should be rejected."""
        raw = bytearray(encode_request(1, PROTO_CMD_POLL, b""))
        raw[0] = 0xFF
        with pytest.raises(ValueError, match="bad START"):
            decode_frame(bytes(raw))

    def test_error_bad_crc(self):
        """Corrupted CRC should be detected."""
        raw = bytearray(encode_request(1, PROTO_CMD_POLL, b""))
        raw[-1] ^= 0xFF
        with pytest.raises(ValueError, match="CRC mismatch"):
            decode_frame(bytes(raw))

    def test_error_length_mismatch_too_long(self):
        """Extra bytes beyond what LEN declares should be rejected."""
        raw = encode_request(1, PROTO_CMD_POLL, b"") + b"\x00"
        with pytest.raises(ValueError, match="length mismatch"):
            decode_frame(raw)

    def test_error_length_mismatch_too_short(self):
        """Frame truncated relative to LEN should be rejected."""
        raw = bytearray(encode_request(1, PROTO_CMD_REPLY, b"\x01\x02\x03"))
        # Truncate one payload byte but keep original LEN
        truncated = bytes(raw[:6])
        with pytest.raises(ValueError):
            decode_frame(truncated)

    def test_error_addr_zero(self):
        """Address 0 in a frame should be rejected.

        We craft a frame with addr=0 and valid CRC to specifically
        test the addr range check.
        """
        import struct
        body = bytes([0x00, PROTO_CMD_POLL, 0x00])
        crc = crc16_modbus(body)
        raw = bytes([PROTO_START]) + body + struct.pack("<H", crc)
        with pytest.raises(ValueError, match="addr out of range"):
            decode_frame(raw)


# -- parse_reply -------------------------------------------------------------


class TestParseReply:
    """Tests for REPLY payload parsing."""

    def test_example2_payload(self):
        """Parse the Example 2 payload: channels 0,1 valid."""
        payload = bytes([
            0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
        ])
        result = parse_reply(payload)
        assert result["temperatures"][0] == pytest.approx(23.5)
        assert result["temperatures"][1] == pytest.approx(19.8)
        assert result["temperatures"][2] is None
        assert result["temperatures"][3] is None

    def test_all_channels_invalid(self):
        """All temps INVALID means all temperatures are None."""
        payload = bytes([0xFF, 0x7F] * 4)
        result = parse_reply(payload)
        assert result["temperatures"] == [None, None, None, None]

    def test_all_channels_valid(self):
        """All 4 temperatures returned as floats."""
        import struct
        temps_raw = struct.pack("<hhhh", 100, -50, 0, 325)
        result = parse_reply(temps_raw)
        assert result["temperatures"][0] == pytest.approx(10.0)
        assert result["temperatures"][1] == pytest.approx(-5.0)
        assert result["temperatures"][2] == pytest.approx(0.0)
        assert result["temperatures"][3] == pytest.approx(32.5)

    def test_negative_temperature(self):
        """Negative temperature values are handled correctly."""
        import struct
        temps_raw = struct.pack("<hhhh", -100, 0, 0, 0)
        result = parse_reply(temps_raw)
        assert result["temperatures"][0] == pytest.approx(-10.0)

    def test_bad_length_short(self):
        """Payload shorter than 8 bytes should be rejected."""
        with pytest.raises(ValueError, match="8 bytes"):
            parse_reply(b"\x00" * 7)

    def test_bad_length_long(self):
        """Payload longer than 8 bytes should be rejected."""
        with pytest.raises(ValueError, match="8 bytes"):
            parse_reply(b"\x00" * 9)


# -- Fuzz-ish ----------------------------------------------------------------


class TestFuzz:
    """Random byte sequences must not crash decode_frame."""

    def test_random_bytes_no_crash(self):
        """Feed 1000 random byte sequences; all must raise ValueError."""
        rng = random.Random(42)
        for _ in range(1000):
            length = rng.randint(0, 30)
            data = bytes(rng.randint(0, 255) for _ in range(length))
            try:
                decode_frame(data)
            except ValueError:
                pass
            # If it doesn't raise, that's fine too -- could be a valid frame

File: sensor/test/test_protocol/test_protocol.cpp
/*
 * test_protocol.c -- Unity tests for tmon protocol encode/decode.
 *
 * Mirrors the Python test coverage in server/tests/test_protocol.py.
 * Run with: pio test -e native
 */

#include <string.h>
#include <unity.h>

#include "protocol.h"

/* -- encode_request tests ------------------------------------------------- */

void
test_encode_example1_poll_sensor3 (void)
{
  /* POLL for sensor 3 should produce the Example 1 frame. */
  uint8_t expected[] = {0x01, 0x03, 0x01, 0x00, 0x80, 0x50};
  uint8_t buf[64];
  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 3, TMON_CMD_POLL,
                                  NULL, 0);
  TEST_ASSERT_EQUAL (6, n);
  TEST_ASSERT_EQUAL_HEX8_ARRAY (expected, buf, 6);
}

void
test_encode_example2_reply_frame (void)
{
  /* encode_request should produce the Example 2 frame. */
  uint8_t payload[] = {
    0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
  };
  uint8_t expected[] = {
    0x01, 0x03, 0x02, 0x08,
    0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
    0x90, 0xEB,
  };
  uint8_t buf[64];
  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 3, TMON_CMD_REPLY,
                                  payload, sizeof (payload));
  TEST_ASSERT_EQUAL (14, n);
  TEST_ASSERT_EQUAL_HEX8_ARRAY (expected, buf, 14);
}

void
test_encode_poll_length_is_6 (void)
{
  /* POLL frame with no payload is always 6 bytes. */
  uint8_t buf[64];
  TEST_ASSERT_EQUAL (6, tmon_proto_encode_frame (buf, sizeof (buf), 1,
                                             TMON_CMD_POLL, NULL, 0));
  TEST_ASSERT_EQUAL (6, tmon_proto_encode_frame (buf, sizeof (buf), 247,
                                             TMON_CMD_POLL, NULL, 0));
}

void
test_encode_start_byte (void)
{
  /* First byte of any frame is the START byte (0x01). */
  uint8_t buf[64];
  tmon_proto_encode_frame (buf, sizeof (buf), 1, TMON_CMD_POLL, NULL, 0);
  TEST_ASSERT_EQUAL_HEX8 (0x01, buf[0]);
}

void
test_encode_addr_in_frame (void)
{
  /* Address appears at offset 1. */
  uint8_t buf[64];
  tmon_proto_encode_frame (buf, sizeof (buf), 42, TMON_CMD_POLL, NULL, 0);
  TEST_ASSERT_EQUAL (42, buf[1]);
}

void
test_encode_cmd_in_frame (void)
{
  /* Command byte appears at offset 2. */
  uint8_t buf[64];
  tmon_proto_encode_frame (buf, sizeof (buf), 1, 0xAB, NULL, 0);
  TEST_ASSERT_EQUAL_HEX8 (0xAB, buf[2]);
}

void
test_encode_len_field (void)
{
  /* LEN field at offset 3 reflects actual payload length. */
  uint8_t buf[64];
  uint8_t payload[] = {0x01, 0x02, 0x03};
  tmon_proto_encode_frame (buf, sizeof (buf), 1, TMON_CMD_POLL,
                       payload, sizeof (payload));
  TEST_ASSERT_EQUAL (3, buf[3]);
}

void
test_encode_addr_zero_rejected (void)
{
  /* Address 0 should be rejected (returns 0). */
  uint8_t buf[64];
  TEST_ASSERT_EQUAL (0, tmon_proto_encode_frame (buf, sizeof (buf), 0,
                                             TMON_CMD_POLL, NULL, 0));
}

void
test_encode_addr_248_rejected (void)
{
  /* Address 248 should be rejected (returns 0). */
  uint8_t buf[64];
  TEST_ASSERT_EQUAL (0, tmon_proto_encode_frame (buf, sizeof (buf), 248,
                                             TMON_CMD_POLL, NULL, 0));
}

void
test_encode_buffer_too_small (void)
{
  /* Buffer too small should return 0. */
  uint8_t buf[4];
  TEST_ASSERT_EQUAL (0, tmon_proto_encode_frame (buf, sizeof (buf), 1,
                                             TMON_CMD_POLL, NULL, 0));
}

/* -- decode_frame tests --------------------------------------------------- */

void
test_decode_roundtrip_poll (void)
{
  /* encode then decode should recover the original fields. */
  uint8_t buf[64];
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 5, TMON_CMD_POLL,
                                  NULL, 0);
  int rc = tmon_proto_decode_frame (buf, n, &addr, &cmd, &payload, &plen);
  TEST_ASSERT_EQUAL (0, rc);
  TEST_ASSERT_EQUAL (5, addr);
  TEST_ASSERT_EQUAL (TMON_CMD_POLL, cmd);
  TEST_ASSERT_EQUAL (0, plen);
  TEST_ASSERT_NULL (payload);
}

void
test_decode_roundtrip_reply (void)
{
  /* Round-trip with a non-empty payload. */
  uint8_t pl[] = {0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F};
  uint8_t buf[64];
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 10, TMON_CMD_REPLY,
                                  pl, sizeof (pl));
  int rc = tmon_proto_decode_frame (buf, n, &addr, &cmd, &payload, &plen);
  TEST_ASSERT_EQUAL (0, rc);
  TEST_ASSERT_EQUAL (10, addr);
  TEST_ASSERT_EQUAL (TMON_CMD_REPLY, cmd);
  TEST_ASSERT_EQUAL (8, plen);
  TEST_ASSERT_EQUAL_HEX8_ARRAY (pl, payload, 8);
}

void
test_decode_example1_from_spec (void)
{
  /* Decode the Example 1 frame from the protocol spec. */
  uint8_t raw[] = {0x01, 0x03, 0x01, 0x00, 0x80, 0x50};
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  int rc = tmon_proto_decode_frame (raw, sizeof (raw), &addr, &cmd,
                              &payload, &plen);
  TEST_ASSERT_EQUAL (0, rc);
  TEST_ASSERT_EQUAL (3, addr);
  TEST_ASSERT_EQUAL (TMON_CMD_POLL, cmd);
  TEST_ASSERT_EQUAL (0, plen);
}

void
test_decode_example2_from_spec (void)
{
  /* Decode the Example 2 frame from the protocol spec. */
  uint8_t raw[] = {
    0x01, 0x03, 0x02, 0x08,
    0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F,
    0x90, 0xEB,
  };
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  int rc = tmon_proto_decode_frame (raw, sizeof (raw), &addr, &cmd,
                              &payload, &plen);
  TEST_ASSERT_EQUAL (0, rc);
  TEST_ASSERT_EQUAL (3, addr);
  TEST_ASSERT_EQUAL (TMON_CMD_REPLY, cmd);
  TEST_ASSERT_EQUAL (8, plen);
}

void
test_decode_error_short_frame (void)
{
  /* Frames shorter than 6 bytes should be rejected. */
  uint8_t raw[] = {0x01, 0x02, 0x03};
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  TEST_ASSERT_EQUAL (-1, tmon_proto_decode_frame (raw, sizeof (raw), &addr, &cmd,
                                            &payload, &plen));
}

void
test_decode_error_bad_start (void)
{
  /* Non-0x01 START byte should be rejected. */
  uint8_t buf[64];
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 1, TMON_CMD_POLL,
                                  NULL, 0);
  buf[0] = 0xFF;
  TEST_ASSERT_EQUAL (-1, tmon_proto_decode_frame (buf, n, &addr, &cmd,
                                            &payload, &plen));
}

void
test_decode_error_bad_crc (void)
{
  /* Corrupted CRC should be detected. */
  uint8_t buf[64];
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 1, TMON_CMD_POLL,
                                  NULL, 0);
  buf[n - 1] ^= 0xFF;
  TEST_ASSERT_EQUAL (-1, tmon_proto_decode_frame (buf, n, &addr, &cmd,
                                            &payload, &plen));
}

void
test_decode_error_length_mismatch_too_long (void)
{
  /* Extra bytes beyond what LEN declares should be rejected. */
  uint8_t buf[64];
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  size_t n = tmon_proto_encode_frame (buf, sizeof (buf), 1, TMON_CMD_POLL,
                                  NULL, 0);
  /* Pass one extra byte */
  TEST_ASSERT_EQUAL (-1, tmon_proto_decode_frame (buf, n + 1, &addr, &cmd,
                                            &payload, &plen));
}

void
test_decode_error_length_mismatch_too_short (void)
{
  /* Frame truncated relative to LEN should be rejected. */
  uint8_t pl[] = {0x01, 0x02, 0x03};
  uint8_t buf[64];
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  tmon_proto_encode_frame (buf, sizeof (buf), 1, TMON_CMD_REPLY,
                       pl, sizeof (pl));
  /* Pass truncated length (6 instead of 9) */
  TEST_ASSERT_EQUAL (-1, tmon_proto_decode_frame (buf, 6, &addr, &cmd,
                                            &payload, &plen));
}

void
test_decode_error_addr_zero (void)
{
  /* Address 0 in a frame should be rejected.
   * Hand-crafted frame: START=0x01, addr=0, cmd=POLL, len=0, CRC=0x5070. */
  uint8_t raw[] = {0x01, 0x00, TMON_CMD_POLL, 0x00, 0x70, 0x50};
  uint8_t addr, cmd, plen;
  const uint8_t *payload;

  TEST_ASSERT_EQUAL (-1, tmon_proto_decode_frame (raw, sizeof (raw), &addr, &cmd,
                                            &payload, &plen));
}

/* -- parse_reply tests ---------------------------------------------------- */

void
test_parse_example2_payload (void)
{
  /* Parse the Example 2 payload: channels 0,1 valid, 2,3 invalid. */
  uint8_t pl[] = {0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F};
  struct tmon_proto_reply_payload rp;

  TEST_ASSERT_EQUAL (0, tmon_proto_parse_reply (pl, sizeof (pl), &rp));
  TEST_ASSERT_EQUAL_INT16 (235, rp.temps[0]);
  TEST_ASSERT_EQUAL_INT16 (198, rp.temps[1]);
  TEST_ASSERT_EQUAL_INT16 ((int16_t)TMON_TEMP_INVALID, rp.temps[2]);
  TEST_ASSERT_EQUAL_INT16 ((int16_t)TMON_TEMP_INVALID, rp.temps[3]);
}

void
test_parse_all_channels_invalid (void)
{
  /* All temps are TMON_TEMP_INVALID. */
  uint8_t pl[] = {0xFF, 0x7F, 0xFF, 0x7F, 0xFF, 0x7F, 0xFF, 0x7F};
  struct tmon_proto_reply_payload rp;

  TEST_ASSERT_EQUAL (0, tmon_proto_parse_reply (pl, sizeof (pl), &rp));
  int i;
  for (i = 0; i < TMON_NUM_CHANNELS; i++)
    TEST_ASSERT_EQUAL_INT16 ((int16_t)TMON_TEMP_INVALID, rp.temps[i]);
}

void
test_parse_all_channels_valid (void)
{
  /* All 4 temperatures returned. */
  /* temps: 100, -50, 0, 325 (LE) */
  uint8_t pl[] = {
    0x64, 0x00,  /* 100 */
    0xCE, 0xFF,  /* -50 */
    0x00, 0x00,  /*   0 */
    0x45, 0x01,  /* 325 */
  };
  struct tmon_proto_reply_payload rp;

  TEST_ASSERT_EQUAL (0, tmon_proto_parse_reply (pl, sizeof (pl), &rp));
  TEST_ASSERT_EQUAL_INT16 (100, rp.temps[0]);
  TEST_ASSERT_EQUAL_INT16 (-50, rp.temps[1]);
  TEST_ASSERT_EQUAL_INT16 (0, rp.temps[2]);
  TEST_ASSERT_EQUAL_INT16 (325, rp.temps[3]);
}

void
test_parse_negative_temperature (void)
{
  /* Negative temperature values are handled correctly. */
  /* -100 in LE = 0x9C, 0xFF */
  uint8_t pl[] = {
    0x9C, 0xFF,  /* -100 */
    0x00, 0x00,
    0x00, 0x00,
    0x00, 0x00,
  };
  struct tmon_proto_reply_payload rp;

  TEST_ASSERT_EQUAL (0, tmon_proto_parse_reply (pl, sizeof (pl), &rp));
  TEST_ASSERT_EQUAL_INT16 (-100, rp.temps[0]);
}

void
test_parse_wrong_length_short (void)
{
  /* Payload shorter than 8 bytes should be rejected. */
  uint8_t pl[7];
  struct tmon_proto_reply_payload rp;
  memset (pl, 0, sizeof (pl));
  TEST_ASSERT_EQUAL (-1, tmon_proto_parse_reply (pl, 7, &rp));
}

void
test_parse_wrong_length_long (void)
{
  /* Payload longer than 8 bytes should be rejected. */
  uint8_t pl[9];
  struct tmon_proto_reply_payload rp;
  memset (pl, 0, sizeof (pl));
  TEST_ASSERT_EQUAL (-1, tmon_proto_parse_reply (pl, 9, &rp));
}

/* -- build_reply_payload tests --------------------------------------------- */

void
test_build_reply_roundtrip (void)
{
  /* Build payload then parse it; temps should match. */
  int16_t input[TMON_NUM_CHANNELS] = {235, 198, 0, 325};
  uint8_t payload[TMON_REPLY_PAYLOAD_LEN];
  struct tmon_proto_reply_payload parsed;
  int i;

  tmon_proto_build_reply_payload (payload, input);
  TEST_ASSERT_EQUAL (0, tmon_proto_parse_reply (payload, TMON_REPLY_PAYLOAD_LEN,
                                          &parsed));
  for (i = 0; i < TMON_NUM_CHANNELS; i++)
    TEST_ASSERT_EQUAL_INT16 (input[i], parsed.temps[i]);
}

void
test_build_reply_known_values (void)
{
  /* Example 2 values: 235, 198, INVALID, INVALID -> known bytes. */
  int16_t temps[TMON_NUM_CHANNELS] = {
    235, 198, (int16_t)TMON_TEMP_INVALID, (int16_t)TMON_TEMP_INVALID
  };
  uint8_t expected[] = {
    0xEB, 0x00, 0xC6, 0x00, 0xFF, 0x7F, 0xFF, 0x7F
  };
  uint8_t payload[TMON_REPLY_PAYLOAD_LEN];

  tmon_proto_build_reply_payload (payload, temps);
  TEST_ASSERT_EQUAL_HEX8_ARRAY (expected, payload, TMON_REPLY_PAYLOAD_LEN);
}

void
test_build_reply_negative_temps (void)
{
  /* Negative values should encode correctly as two's complement LE. */
  int16_t temps[TMON_NUM_CHANNELS] = {-100, -50, -1, 0};
  uint8_t payload[TMON_REPLY_PAYLOAD_LEN];
  struct tmon_proto_reply_payload parsed;
  int i;

  tmon_proto_build_reply_payload (payload, temps);

  /* -100 = 0xFF9C -> LE: 0x9C, 0xFF */
  TEST_ASSERT_EQUAL_HEX8 (0x9C, payload[0]);
  TEST_ASSERT_EQUAL_HEX8 (0xFF, payload[1]);
  /* -50 = 0xFFCE -> LE: 0xCE, 0xFF */
  TEST_ASSERT_EQUAL_HEX8 (0xCE, payload[2]);
  TEST_ASSERT_EQUAL_HEX8 (0xFF, payload[3]);
  /* -1 = 0xFFFF -> LE: 0xFF, 0xFF */
  TEST_ASSERT_EQUAL_HEX8 (0xFF, payload[4]);
  TEST_ASSERT_EQUAL_HEX8 (0xFF, payload[5]);

  /* Roundtrip check */
  TEST_ASSERT_EQUAL (0, tmon_proto_parse_reply (payload, TMON_REPLY_PAYLOAD_LEN,
                                          &parsed));
  for (i = 0; i < TMON_NUM_CHANNELS; i++)
    TEST_ASSERT_EQUAL_INT16 (temps[i], parsed.temps[i]);
}

/* -- Unity setup/teardown ------------------------------------------------- */

void
setUp (void)
{
}

void
tearDown (void)
{
}

int
main (void)
{
  UNITY_BEGIN ();

  /* Encode */
  RUN_TEST (test_encode_example1_poll_sensor3);
  RUN_TEST (test_encode_example2_reply_frame);
  RUN_TEST (test_encode_poll_length_is_6);
  RUN_TEST (test_encode_start_byte);
  RUN_TEST (test_encode_addr_in_frame);
  RUN_TEST (test_encode_cmd_in_frame);
  RUN_TEST (test_encode_len_field);
  RUN_TEST (test_encode_addr_zero_rejected);
  RUN_TEST (test_encode_addr_248_rejected);
  RUN_TEST (test_encode_buffer_too_small);

  /* Decode */
  RUN_TEST (test_decode_roundtrip_poll);
  RUN_TEST (test_decode_roundtrip_reply);
  RUN_TEST (test_decode_example1_from_spec);
  RUN_TEST (test_decode_example2_from_spec);
  RUN_TEST (test_decode_error_short_frame);
  RUN_TEST (test_decode_error_bad_start);
  RUN_TEST (test_decode_error_bad_crc);
  RUN_TEST (test_decode_error_length_mismatch_too_long);
  RUN_TEST (test_decode_error_length_mismatch_too_short);
  RUN_TEST (test_decode_error_addr_zero);

  /* Parse reply */
  RUN_TEST (test_parse_example2_payload);
  RUN_TEST (test_parse_all_channels_invalid);
  RUN_TEST (test_parse_all_channels_valid);
  RUN_TEST (test_parse_negative_temperature);
  RUN_TEST (test_parse_wrong_length_short);
  RUN_TEST (test_parse_wrong_length_long);

  /* Build reply payload */
  RUN_TEST (test_build_reply_roundtrip);
  RUN_TEST (test_build_reply_known_values);
  RUN_TEST (test_build_reply_negative_temps);

  return UNITY_END ();
}
